File: wheels/wheels.c
#include <stdlib.h>
#include <kipr/wombat.h>
#include "wheels.h"

double scale(double n, double in_min, double in_max, double out_min, double out_max) {
    return (n - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
}

void drive_motor(Motor motor, DriveDirection direction, double cm) {
    int velocity = motor.speed * MOTOR_PWM_TICKS;

    if (direction == REVERSE) {
        velocity *= -1;
    }

    move_at_velocity(motor.port, velocity);

    Duration block_duration = motor_block_duration(cm, velocity);

    msleep((int) block_duration);

    force_stop_motor(motor);
}

void force_stop_motor(Motor motor) {
    move_at_velocity(motor.port, 0);
    msleep(50);
    off(motor.port);
}

Duration motor_block_duration(double cm, int velocity) {
    return fabs(MOTOR_PWM_TICKS * MOTOR_TRAVEL_TIME_1_CM * cm / (double) velocity) * 1000.0;
}

void drive_wheels(Wheels wheels, DriveDirection direction, double cm) {
    int left_velocity = calculate_velocity(wheels.left_motor.speed, direction, wheels.left_offset);
    int right_velocity = calculate_velocity(wheels.right_motor.speed, direction, wheels.right_offset);

    int slower_velocity = left_velocity < right_velocity ? left_velocity : right_velocity;
    int sleep_time = motor_block_duration(cm, slower_velocity);

    drive_in_unison(wheels, left_velocity, right_velocity, sleep_time);
}

void turn_wheels(Wheels wheels, TurnDirection direction, double deg) {
    DriveDirection left_direction;
    DriveDirection right_direction;

    if (direction == LEFT) {
        left_direction = REVERSE;
        right_direction = FORWARD;
    } else {
        left_direction = FORWARD;
        right_direction = REVERSE;
    }

    int left_velocity = calculate_velocity(wheels.left_motor.speed, left_direction, wheels.left_offset);
    int right_velocity = calculate_velocity(wheels.right_motor.speed, right_direction, wheels.right_offset);

    double cm = wheels_turn_amount(deg);
    Duration sleep_time = motor_block_duration(cm, left_velocity);

    drive_in_unison(wheels, left_velocity, right_velocity, sleep_time);
}

void drive_in_unison(Wheels wheels, int left_velocity, int right_velocity, Duration sleep_time) {
    move_at_velocity(wheels.left_motor.port, left_velocity);
    move_at_velocity(wheels.right_motor.port, right_velocity);

    msleep(sleep_time);

    force_stop_wheels(wheels);
}

void force_stop_wheels(Wheels wheels) {
    move_at_velocity(wheels.left_motor.port, 0);
    move_at_velocity(wheels.right_motor.port, 0);
    msleep(50);
    off(wheels.left_motor.port);
    off(wheels.right_motor.port);
}

int calculate_velocity(double speed, DriveDirection direction, double offset) {
    int velocity = (int) (speed * MOTOR_PWM_TICKS * offset);

    if (direction == REVERSE) {
        velocity *= -1;
    }

    return velocity;
}

double wheels_turn_amount(double degrees) {
    int multiplier;

    if (degrees == 45) {
        multiplier = 1.175;
    } else if (degrees == 90) {
        multiplier = 1.2425;
    } else if (degrees == 180) {
        multiplier = 1.35;
    } else if (degrees == 360) {
        multiplier = 1.425;
    } else {
        multiplier =
            -2.781096509 * powf(10.0, -6.0) * powf(degrees, 2.0)
                + 1.922759857 * powf(10.0, -3.0) * degrees
                + 1.093333333;
    }

    return degrees * multiplier / 10.0;
}

void set_servo(Servo servo, int position) {
    enable_servos();
    set_servo_position(servo.port, position);
    msleep(750);
}

File: src/lib/wheels.py
from enum import Enum
import math
import sys

from .libwallaby import libwallaby

motor_pwm_ticks = 1500
motor_travel_time_1_cm = 0.0625


def m(x):
    return x * 100


def cm(x):
    return x


def mm(x):
    return x / 10


def inches(x):
    return x * 2.54


class Direction(Enum):
    forward = 0
    reverse = 1

    def toggle(self):
        return Direction.reverse if self == Direction.forward else Direction.forward


class TurnDirection(Enum):
    left = 0
    right = 1

    def toggle(self):
        return TurnDirection.right if self == TurnDirection.left else TurnDirection.left


def scale(n, in_min, in_max, out_min, out_max):
    return (n - in_min) * (out_max - out_min) / (in_max - in_min) + out_min


class Motor:
    def __init__(self, port, speed):
        self.port = port
        self.speed = speed

    def drive(self, direction, cm):
        velocity = int(self.speed * motor_pwm_ticks)

        if direction == Direction.reverse:
            velocity *= -1

        libwallaby.move_at_velocity(self.port, velocity)

        block_duration = motor_block_duration(cm, velocity)

        libwallaby.msleep(int(block_duration))

        self.stop()

    def start(self, direction):
        velocity = int(self.speed * motor_pwm_ticks)

        if direction == Direction.reverse:
            velocity *= -1

        libwallaby.move_at_velocity(self.port, velocity)

    def stop(self):
        libwallaby.move_at_velocity(self.port, 0)
        libwallaby.msleep(50)
        libwallaby.off(self.port)


def motor_block_duration(cm, velocity):
    return int(abs(motor_pwm_ticks * motor_travel_time_1_cm * cm / velocity) * 1000)


class Wheels:
    def __init__(self, left_motor, right_motor, left_offset=1, right_offset=1):
        self.left_motor = left_motor
        self.right_motor = right_motor
        self.left_offset = left_offset
        self.right_offset = right_offset

    def drive(self, direction, cm):
        left_velocity = calculate_velocity(
            self.left_motor.speed, direction, self.left_offset
        )
        right_velocity = calculate_velocity(
            self.right_motor.speed, direction, self.right_offset
        )

        slower_velocity = (
            left_velocity if left_velocity < right_velocity else right_velocity
        )
        sleep_time = motor_block_duration(cm, slower_velocity)

        self.drive_in_unison(left_velocity, right_velocity, sleep_time)

    def start_turn(self, direction):
        if direction == TurnDirection.left:
            left_direction = Direction.forward
            right_direction = Direction.reverse
        else:
            left_direction = Direction.reverse
            right_direction = Direction.forward

        left_velocity = calculate_velocity(
            self.left_motor.speed, left_direction, self.left_offset
        )
        right_velocity = calculate_velocity(
            self.right_motor.speed, right_direction, self.right_offset
        )

        libwallaby.move_at_velocity(self.left_motor.port, left_velocity)
        libwallaby.move_at_velocity(self.right_motor.port, right_velocity)

    def turn(self, direction, deg):
        if direction == TurnDirection.left:
            left_direction = Direction.forward
            right_direction = Direction.reverse
        else:
            left_direction = Direction.reverse
            right_direction = Direction.forward

        left_velocity = calculate_velocity(
            self.left_motor.speed, left_direction, self.left_offset
        )
        right_velocity = calculate_velocity(
            self.right_motor.speed, right_direction, self.right_offset
        )

        cm = wheels_turn_amount(deg)
        sleep_time = motor_block_duration(cm, left_velocity)

        self.drive_in_unison(left_velocity, right_velocity, sleep_time)

    def drive_in_unison(self, left_velocity, right_velocity, sleep_time):
        libwallaby.move_at_velocity(self.left_motor.port, left_velocity)
        libwallaby.move_at_velocity(self.right_motor.port, right_velocity)

        libwallaby.msleep(sleep_time)

        self.stop()

    def start(self, direction):
        left_velocity = calculate_velocity(
            self.left_motor.speed, direction, self.left_offset
        )

        right_velocity = calculate_velocity(
            self.right_motor.speed, direction, self.right_offset
        )

        libwallaby.move_at_velocity(self.left_motor.port, left_velocity)
        libwallaby.move_at_velocity(self.right_motor.port, right_velocity)

    def stop(self):
        libwallaby.move_at_velocity(self.left_motor.port, 0)
        libwallaby.move_at_velocity(self.right_motor.port, 0)
        libwallaby.msleep(50)
        libwallaby.off(self.left_motor.port)
        libwallaby.off(self.right_motor.port)


def calculate_velocity(speed, direction, offset):
    velocity = int(speed * motor_pwm_ticks * offset)

    if direction == Direction.reverse:
        velocity *= -1

    return velocity


def wheels_turn_amount(degrees):
    if degrees == 45:
        multiplier = 1.175
    elif degrees == 90:
        multiplier = 1.2425
    elif degrees == 180:
        multiplier = 1.35
    elif degrees == 360:
        multiplier == 1.425
    else:
        multiplier = (
            -2.781096509 * 10.0**-6.0 * degrees**2.0
            + 1.922759857 * 10.0**-3.0 * degrees
            + 1.093333333
        )

    return degrees * multiplier / 10


class Servo:
    def __init__(self, port, delay):
        self.port = port
        self.delay = delay

    def set(self, target_position):
        assert self.delay > 0

        libwallaby.enable_servos()

        position = libwallaby.get_servo_position(self.port)
        sign = int(math.copysign(1, target_position - position))
        increment = 5 * sign

        while abs(position - target_position) >= 5:
            position += increment
            libwallaby.set_servo_position(self.port, position)
            libwallaby.msleep(self.delay)
